grails-app/services/org/pih/warehouse/putaway/PutawayTaskService.groovy

```diff
@@ -267,6 +285,81 @@ class PutawayTaskService {
         save(task)
     }
 
+    PutawayTask partialComplete(PutawayTask task, BigDecimal quantity, String destinationId, Boolean force = false) {
+        if (quantity > task.quantity) {
+            throw new IllegalArgumentException("Quantity provided is more than requested. Please re-enter quantity")
+        }
+
+        BigDecimal quantityRemaining = task.quantity - quantity
+        if (quantityRemaining <= 0) {
+            throw new IllegalArgumentException("Quantity provided is more than requested. Please re-enter quantity")
+        }
+
+        // validate destination or user has forced a destination change
+        Location destination = Location.get(destinationId)
+        if (!destination) {
+            task.errors.reject("destination", "Destination is required")
+        }
+
+        // validate destination
+        if (task.destination != destination && !force) {
+            task.errors.reject("destination", "Destination provided does not match expected destination")
+        }
+
+        // Update the task destination if the destinations do not match, but user forced change
+        if (task.destination != destination && force) {
+            task.destination = destination
+        }
+
+        if (task.hasErrors() || !task.validate()) {
+            throw new ValidationException("Validation errors occurred during complete action", task.errors)
+        }
+
+        OrderItem currentItem = OrderItem.get(task.putawayOrderItem.id)
+        Order order = currentItem.order
+        OrderItem originalItem = order.orderItems.find { it.parentOrderItem == null } as OrderItem
+
+        Putaway putaway = Putaway.createFromOrder(order)
+        if (!putaway.putawayAssignee) {
+            putaway.putawayAssignee = AuthService.currentUser
+        }
+        PutawayItem itemToSplit = putaway.putawayItems.find { it.id == originalItem.id}
+
+        PutawayItem completedSplitItem = createSplitPutawayItem(task, quantity, PutawayStatus.COMPLETED)
+        PutawayItem remainingSplitItem = createSplitPutawayItem(task, quantityRemaining, PutawayStatus.PENDING)
+
+        if (itemToSplit) {
+            itemToSplit.splitItems = [completedSplitItem, remainingSplitItem]
+        }
+        putawayService.savePutaway(putaway)
+        currentItem.orderItemStatusCode = OrderItemStatusCode.CANCELED
+        currentItem.save(flush: true, failOnError: true)
+
+        OrderItem remainingOrderItem = order.orderItems.find {
+            it.parentOrderItem?.id == originalItem.id &&
+                    it.quantity == quantityRemaining &&
+                    it.orderItemStatusCode == OrderItemStatusCode.PENDING
+        } as OrderItem
+
+        return PutawayTask.createFromOrderItem(remainingOrderItem)
+    }
+
+    void shortage(PutawayTask task, ReasonCode reasonCode) {
+        def putawayDiscrepancyLocation = task.facility.internalLocations
+            .find { it.supports(ActivityCode.PUTAWAY_DISCREPANCY) }
```

Comment:
We should throw an exception here (or implicitly create it) if the putaway discrepancy location doesn't exist.